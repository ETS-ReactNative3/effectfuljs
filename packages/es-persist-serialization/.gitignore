index.js
node_modules
tmp-*
*.orig
*.log
*~
*.map
.DS_Store
.tern-port
err-log.json
