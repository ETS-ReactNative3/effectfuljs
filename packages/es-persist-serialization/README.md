# @effectful/es-persist-serialization

The library uses [@effectful/es-persist](https://github.com/awto/effectfuljs/tree/master/packages/serialization)
to store async function state for
[@effectful/es-persist](https://github.com/awto/effectfuljs/tree/master/packages/es-persist).

It re-exports both of the libraries, and `@effectful/serialization` is
imported as peer dependency.

## Usage

Install from npm:

```
$ npm install --save-dev @effectful/es
$ npm install --save @effectful/es-persist-serialization @effectful/serialization
```

In .babelrc:

```json
{
  "presets": ["@effectful/es-persist-serialization/transform"]
}

```

If there are other babel presets, add `"passPerPreset":true`.

## License

Distributed under the terms of the [The MIT License (MIT)](LICENSE). 

