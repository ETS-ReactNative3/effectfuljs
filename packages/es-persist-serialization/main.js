import * as S from "@effectful/serialization";
import * as R from "@effectful/es-persist";
export {
  generator,
  generatorFunction,
  iterator,
  iteratorM,
  forInIterator,
  contextSymbol,
  sendSymbol,
  awaitSymbol,
  context,
  async,
  asyncGenerator,
  lock,
  all,
  any,
  producer,
  current,
  idle,
  abort,
  managed
} from "@effectful/es-persist";

export {
  read,
  write,
  regConstructor,
  regOpaqueObject,
  bind,
  descriptorSymbol,
  parse,
  stringify,
  regDescriptor,
  regOpaquePrim
} from "@effectful/serialization";

R.constructors.forEach(v => S.regConstructor(v));
export const asyncFunction = R.makeAsyncFunctionConstructor(S.regConstructor);
export const asyncGeneratorFunction = R.makeAsyncGeneratorFunctionConstructor(
  S.regConstructor
);
