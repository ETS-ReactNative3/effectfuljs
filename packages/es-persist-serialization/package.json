{
  "name": "@effectful/es-persist-serialization",
  "version": "1.3.6",
  "description": "@effectful/es-persist with @effectful/serialization for storing state",
  "main": "index.js",
  "module": "main.js",
  "jsnext:main": "main.js",
  "scripts": {
    "test": "mocha",
    "build": "babel --config-file ../es/babel.cjs.json -o index.js main.js"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/awto/effectfuljs.git"
  },
  "keywords": [
    "effectful",
    "effectfuljs",
    "monad",
    "linq"
  ],
  "author": "Vitaliy Akimov <vitaliy.akimov@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/awto/effectfuljs/issues"
  },
  "homepage": "https://github.com/awto/effectfuljs#readme",
  "peerDependencies": {
    "@effectful/serialization": "^1.0.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.0.0",
    "@effectful/serialization": "^1.3.6",
    "mocha": "^5.0.5"
  },
  "dependencies": {
    "@effectful/es-persist": "^1.3.1"
  },
  "gitHead": "d3b25d1148c0b8da0bb45a6a52b8b8cac3150ca5"
}
