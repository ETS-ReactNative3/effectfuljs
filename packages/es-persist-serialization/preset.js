module.exports = require("@effectful/es-persist/preset")
  .options({importRT:"@effectful/es-persist-serialization"})
