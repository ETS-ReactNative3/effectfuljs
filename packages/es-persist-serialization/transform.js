module.exports = require("@effectful/es-persist/transform").options({
  importRT: "@effectful/es-persist-serialization"
});
