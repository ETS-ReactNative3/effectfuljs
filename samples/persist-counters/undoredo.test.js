import * as R from "@effectful/es-persist-serialization"
import undoRedo from "./undoredo"


test("undoRedo", () => {
  let cb, ce
  const exit = { resume() { cb() }, reject(e) { ce(e) } }
  R.regOpaqueObject(exit);
  async function* value(input) {
    let cnt = 0
    for await(const i of input) {
      if (i.type === "INCREMENT")
        cnt++
      else if (i.type === "FLUSH")
        yield {type:"VALUE",value:cnt}
      yield i
    }
  }
  R.regOpaqueObject(value)
  const producer = R.producer()
  R.regOpaqueObject(producer,"producer")
  let deb_num = 0
  for(const type of ["INCREMENT","INCREMENT","INCREMENT",
                     "UNDO","UNDO","REDO","REDO",
                     "UNDO","INCREMENT","UNDO"
                    ]) {
    producer.send({type,x:deb_num++})
    producer.send({type:"FLUSH",x:deb_num++})
  }
  producer.stop();
  (async function main()  {
    if (!await R.managed)
      return
    for await(const i of value(undoRedo(producer))) {
      if (i.type === "VALUE" || i.type === "MENU_ITEM")
        expect(R.write(i)).toMatchSnapshot()
    }
  })()[R.awaitSymbol](exit)
  return new Promise((r, e) => (cb = r, ce = e))

})

