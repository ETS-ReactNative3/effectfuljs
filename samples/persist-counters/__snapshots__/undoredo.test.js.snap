// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`undoRedo 1`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 0,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 0,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "$": "_signal",
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 2`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      1,
    ],
  ],
}
`;

exports[`undoRedo 3`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 0,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 0,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "$": "_signal",
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 4`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      2,
    ],
  ],
}
`;

exports[`undoRedo 5`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 0,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 0,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "$": "_signal",
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 6`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      3,
    ],
  ],
}
`;

exports[`undoRedo 7`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 1,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
          Object {
            "$": "ReactElement",
            "children": Array [
              "Redo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 2,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".1",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "_signal",
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 1,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "REDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 2,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 8`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      2,
    ],
  ],
}
`;

exports[`undoRedo 9`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 1,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
          Object {
            "$": "ReactElement",
            "children": Array [
              "Redo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 2,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".1",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "_signal",
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 1,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "REDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 2,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 10`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      1,
    ],
  ],
}
`;

exports[`undoRedo 11`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 1,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
          Object {
            "$": "ReactElement",
            "children": Array [
              "Redo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 2,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".1",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "_signal",
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 1,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "REDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 2,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 12`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      2,
    ],
  ],
}
`;

exports[`undoRedo 13`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 0,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 0,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "$": "_signal",
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 14`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      3,
    ],
  ],
}
`;

exports[`undoRedo 15`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 1,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
          Object {
            "$": "ReactElement",
            "children": Array [
              "Redo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 2,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".1",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "_signal",
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 1,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "REDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 2,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 16`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      2,
    ],
  ],
}
`;

exports[`undoRedo 17`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 0,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 0,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "$": "_signal",
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 18`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      3,
    ],
  ],
}
`;

exports[`undoRedo 19`] = `
Object {
  "f": Array [
    Array [
      "type",
      "MENU_ITEM",
    ],
    Array [
      "index",
      200,
    ],
    Array [
      "value",
      Object {
        "$": "ReactElement",
        "children": Array [
          Object {
            "$": "ReactElement",
            "children": Array [
              "Undo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 1,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".0",
                ],
              ],
            },
            "type": "button",
          },
          Object {
            "$": "ReactElement",
            "children": Array [
              "Redo",
            ],
            "props": Object {
              "f": Array [
                Array [
                  "onClick",
                  Object {
                    "r": 2,
                  },
                ],
                Array [
                  "children",
                  Object {
                    "$": "undefined",
                  },
                ],
                Array [
                  "key",
                  ".1",
                ],
              ],
            },
            "type": "button",
          },
        ],
        "props": Object {
          "f": Array [
            Array [
              "children",
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        "type": Object {
          "$": "Symbol",
          "key": "react.fragment",
        },
      },
    ],
  ],
  "x": Array [
    Object {
      "$": "_signal",
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "UNDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 1,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
    Object {
      "$": "Bind",
      "args": Array [
        Object {
          "$": "undefined",
        },
        "REDO",
        Object {
          "$": "undefined",
        },
      ],
      "f": Array [
        Array [
          "prototype",
          Object {
            "f": Array [
              Array [
                "constructor",
                Object {
                  "r": 2,
                },
                2,
              ],
            ],
          },
          3,
        ],
      ],
      "func": Object {
        "r": 0,
      },
      "self": null,
    },
  ],
}
`;

exports[`undoRedo 20`] = `
Object {
  "f": Array [
    Array [
      "type",
      "VALUE",
    ],
    Array [
      "value",
      2,
    ],
  ],
}
`;
